Docs/Extensions-Security.md

```diff
@@ -0,0 +1,184 @@
+
+- Defining permissions
+- Defining roles
+- Checking user permissions
+
+### Definite permissions
+
+- TODO How should permissions be defined?
+
+Hypothetical example of permissions:
+
+```yaml
+# permissions.yml
+
+# Example of named permission
+EditDetectorWithConstraints:
+  action: EditDetector           # Action would be defined by an extension developer
+  description: Edit detectors with restrictions
+  constraints:
+    - resource_name: *           # [NamePattern]
+    - index_patterns:
+      - ad*                      # [IndexPattern]
+    - created_by:
+      - user1
+      - user2
+    - created_at: ">=2000-01-01" # [DatetimeExpression]
+
+EditOwnDetector:
+  action: EditDetector
+  description: Edit detector owned by current user
+  constraints:
+    created_by:
+      - "${user_name}"
+  
+# TODO Are some constraints required and others optional?
+```
+
+See the section on `Checking user permissions` to see how the permissions can be checked on the handling of an AuthorizationRequest.
+
+### Defining roles
+
+- TODO How should roles be defined?
+
+Hypothetical example of roles:
+
+```yaml
+# roles.yml
+
+DetectorResultsRO:
+  permissions:
+    - action: CreateDetector
+      description: Create detector
+    - action: ListResults
+      description: List results
+    - action: CreateResults
+      description: Create detector results
+  named_permissions:
+    - EditOwnDetector
+```
+
+This is an example of how a plugin defines roles currently:
+
+```yaml
+# Allow users to read Anomaly Detection detectors and results
+anomaly_read_access:
+  reserved: true
+  cluster_permissions:
+    - 'cluster:admin/opendistro/ad/detector/info'
+    - 'cluster:admin/opendistro/ad/detector/search'
+    - 'cluster:admin/opendistro/ad/detectors/get'
+    - 'cluster:admin/opendistro/ad/result/search'
+    - 'cluster:admin/opendistro/ad/tasks/search'
+    - 'cluster:admin/opendistro/ad/detector/validate'
+    - 'cluster:admin/opendistro/ad/result/topAnomalies'
+
+# Allows users to use all Anomaly Detection functionality
+anomaly_full_access:
+  reserved: true
+  cluster_permissions:
+    - 'cluster_monitor'
+    - 'cluster:admin/opendistro/ad/*'
+  index_permissions:
+    - index_patterns:
+        - '*'
+      allowed_actions:
+        - 'indices_monitor'
+        - 'indices:admin/aliases/get'
+        - 'indices:admin/mappings/get'
+```
+
+### Checking user permissions
+
+Before any action is performed on an extension, the action should be preceded with an `AuthorizationRequest` to 
+determine whether the user of an extension has the required privileges to perform the action. An authorization
+request consists of the name of a permission, a user identifier and a map of required parameters needed to check if a 
+user has the required permissions. AuthorizationRequest handling is performed in core and yet to be implemented.
+
+```java
+public class AuthorizationRequest extends TransportRequest {
```

Comment:
@cwperks Please create a separate PR in OpenSearch for the AuthorizationRequest so we can discuss the design elements and see if we should merge it.

Reply:
@peternied The change I am making relies on @DarshitChanpura's unmerged changes: https://github.com/opensearch-project/OpenSearch/pull/4299/files#diff-a1079c2a0ec10aedc11808ac7129483612a01d759d34bf3499b76cdce5c98cfb

The diff is available here: https://github.com/DarshitChanpura/OpenSearch/compare/principal-identity-extensions...cwperks:OpenSearch:authorize-request

I am not able to create a PR based on the branch by @DarshitChanpura

Reply:
I'll change the requesterIdentifier from an object to a string in order to create a PR and it will be reworked with @DarshitChanpura's branch when that is ready.

Reply:
@peternied Here's the PR for adding `AuthorizationRequest` to `feature/extensions` of core: https://github.com/opensearch-project/OpenSearch/pull/4516